File: Afonso/Serie1/src/main/utils/TimeoutHolder.java
package utils;

import java.util.concurrent.TimeUnit;

public class TimeoutHolder {
    private final long deadline;		// -1L when timeout is infinite

    public TimeoutHolder(long millis) {
        deadline = millis >= 0L ? System.currentTimeMillis() + millis: -1L;
    }

    public TimeoutHolder(long time, TimeUnit unit) {
        deadline = time >= 0L ? System.currentTimeMillis() + unit.toMillis(time) : -1L;
    }

    // returns true if timeout exists
    public boolean isTimed() { return deadline >= 0L; }

    public long value() {
        if (deadline == -1L)
            return Long.MAX_VALUE;	// ensure that timeout does not expire!
        long remainder = deadline - System.currentTimeMillis();
        return Math.max(remainder, 0L);
    }
}

File: Afonso/Serie1 .NET/.NET/Serie1/Utils/TimeoutHolder.cs
using System;
using System.Threading;

namespace Utils {
    public struct TimeoutHolder {
        private int _timeout;
        private int _refTime;

        public TimeoutHolder(int timeout) {
            _timeout = timeout;
            _refTime = (timeout != 0 && timeout != Timeout.Infinite) ? Environment.TickCount : 0;
        }

        // returns the remaining timeout
        public int Value {
            get {
                if (_timeout != 0 && _timeout != Timeout.Infinite) {
                    int now = Environment.TickCount;
                    if (now != _refTime) {
                        int elapsed = now - _refTime;
                        _refTime = now;
                        _timeout = elapsed < _timeout ? _timeout - elapsed : 0;
                    }
                }

                return _timeout;
            }
        }
    }
}